// @injectivelabs/core-proto-ts-v2/generated/injective/permissions/v1beta1/params_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class Params$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.Params", [
      {
        no: 1,
        name: "contract_hook_max_gas",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "enforced_restrictions_contracts",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.equal": true, "amino.name": "permissions/Params" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.contractHookMaxGas = 0n;
    message.enforcedRestrictionsContracts = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 contract_hook_max_gas */
        1:
          message.contractHookMaxGas = reader.uint64().toBigInt();
          break;
        case /* repeated string enforced_restrictions_contracts */
        2:
          message.enforcedRestrictionsContracts.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.contractHookMaxGas !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.contractHookMaxGas);
    for (let i = 0; i < message.enforcedRestrictionsContracts.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.enforcedRestrictionsContracts[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
export {
  Params
};
