// @injectivelabs/indexer-proto-ts-v2/generated/injective_rfqrpc_pb.client.js
import { InjectiveRFQRPC } from "./injective_rfqrpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveRFQRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveRFQRPC.typeName;
    this.methods = InjectiveRFQRPC.methods;
    this.options = InjectiveRFQRPC.options;
  }
  /**
   * Create RFQ request
   *
   * @generated from protobuf rpc: Request
   */
  request(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream RFQ requests
   *
   * @generated from protobuf rpc: StreamRequest
   */
  streamRequest(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Create RFQ quote
   *
   * @generated from protobuf rpc: Quote
   */
  quote(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream RFQ quotes
   *
   * @generated from protobuf rpc: StreamQuote
   */
  streamQuote(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Get open RFQ requests
   *
   * @generated from protobuf rpc: GetOpenRequests
   */
  getOpenRequests(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get pending RFQ quotes
   *
   * @generated from protobuf rpc: GetPendingQuotes
   */
  getPendingQuotes(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List RFQ settlements
   *
   * @generated from protobuf rpc: ListSettlement
   */
  listSettlement(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream RFQ settlements
   *
   * @generated from protobuf rpc: StreamSettlement
   */
  streamSettlement(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Bidirectional stream for takers: send requests, receive quotes
   *
   * @generated from protobuf rpc: TakerStream
   */
  takerStream(options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("duplex", this._transport, method, opt);
  }
  /**
   * Bidirectional stream for makers: receive requests, send quotes
   *
   * @generated from protobuf rpc: MakerStream
   */
  makerStream(options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("duplex", this._transport, method, opt);
  }
}
export {
  InjectiveRFQRPCClient
};
