// @injectivelabs/sdk-ts/dist/cjs/cosmjs.cjs
require('./utils-D3XFGv8y.cjs');
require('./defineProperty-DAtJs5JQ.cjs');
require('./accounts-CAskEh_H.cjs');
require('./grpc-FG2z-ICE.cjs');
require('./MsgBase-CJJkzmQk.cjs');
require('./ChainRestTendermintApi-DILpgVDN.cjs');
require('./BaseGrpcConsumer-B1hpRQJ-.cjs');
require('./BaseIndexerGrpcConsumer-rRYRp48i.cjs');
require('./BaseRestConsumer-CRkyjiSo.cjs');
require('./IndexerGrpcWeb3GwApi-wpbu-sy0.cjs');
require('./tx-jKdvnSF8.cjs');
const require_DirectEthSecp256k1Wallet = require('./DirectEthSecp256k1Wallet-Dspl1Hmi.cjs');

exports.InjectiveDirectEthSecp256k1Wallet = require_DirectEthSecp256k1Wallet.DirectEthSecp256k1Wallet;
exports.InjectiveEthSecp256k1Wallet = require_DirectEthSecp256k1Wallet.EthSecp256k1Wallet;
exports.InjectiveSigningStargateClient = require_DirectEthSecp256k1Wallet.SigningStargateClient;
exports.InjectiveStargateClient = require_DirectEthSecp256k1Wallet.StargateClient;
exports.MsgTransferCosmjs = require_DirectEthSecp256k1Wallet.MsgTransferCosmjs;
exports.injectiveAccountParser = require_DirectEthSecp256k1Wallet.accountParser;
